src/ammo_effect.cpp

```diff
@@ -0,0 +1,129 @@
+#include "ammo_effect.h"
+
+#include "generic_factory.h"
+#include "json.h"
+
+namespace
+{
+
+generic_factory<ammo_effect> all_ammo_effects( "ammo effects" );
+
+} // namespace
+
+/** @relates int_id */
+template<>
+bool int_id<ammo_effect>::is_valid() const
+{
+    return all_ammo_effects.is_valid( *this );
+}
+
+/** @relates int_id */
+template<>
+const ammo_effect &int_id<ammo_effect>::obj() const
+{
+    return all_ammo_effects.obj( *this );
+}
+
+/** @relates int_id */
+template<>
+const string_id<ammo_effect> &int_id<ammo_effect>::id() const
+{
+    return all_ammo_effects.convert( *this );
+}
+
+/** @relates string_id */
+template<>
+bool string_id<ammo_effect>::is_valid() const
+{
+    return all_ammo_effects.is_valid( *this );
+}
+
+/** @relates string_id */
+template<>
+const ammo_effect &string_id<ammo_effect>::obj() const
+{
+    return all_ammo_effects.obj( *this );
+}
+
+/** @relates string_id */
+template<>
+int_id<ammo_effect> string_id<ammo_effect>::id() const
+{
+    return all_ammo_effects.convert( *this, AE_NULL );
+}
+
+/** @relates int_id */
+template<>
+int_id<ammo_effect>::int_id( const string_id<ammo_effect> &id ) : _id( id.id() )
+{
+}
+
+void ammo_effect::load( const JsonObject &jo, const std::string & )
+{
+    if( jo.has_member( "aoe" ) ) {
+        JsonObject joa = jo.get_object( "aoe" );
+        optional( joa, was_loaded, "field_type", aoe_field_type_name, "fd_null" );
+        optional( joa, was_loaded, "intensity_min", aoe_intensity_min, 0 );
+        optional( joa, was_loaded, "intensity_max", aoe_intensity_max, 0 );
+        optional( joa, was_loaded, "radius", aoe_radius, 1 );
+        optional( joa, was_loaded, "radius_z", aoe_radius_z, 0 );
+        optional( joa, was_loaded, "chance", aoe_chance, 1 );
+        optional( joa, was_loaded, "size", aoe_size, 0 );
+        optional( joa, was_loaded, "check_passable", aoe_check_passable, false );
+        optional( joa, was_loaded, "check_sees", aoe_check_sees, false );
+        optional( joa, was_loaded, "check_sees_radius", aoe_check_sees_radius, 0 );
+    }
+    if( jo.has_member( "explosion" ) ) {
+        JsonObject joe = jo.get_object( "explosion" );
+        aoe_explosion_data = load_explosion_data( joe );
+    }
+    optional( jo, was_loaded, "do_flashbang", do_flashbang, false );
+    optional( jo, was_loaded, "do_emp_blast", do_emp_blast, false );
+}
+
+void ammo_effect::finalize()
+{
+    for( const ammo_effect &ae : ammo_effects::get_all() ) {
+        const_cast<ammo_effect &>( ae ).aoe_field_type = field_type_id( ae.aoe_field_type_name );
+    }
+
+}
+
+void ammo_effect::check() const
+{
```

Comment:
You could check `aoe_field_type` to be valid.

Or check that ranges are correct: min value must be less or equal to max value. Or check that certain integer values are not negative (e.g. aoe_check_sees_radius).